### PR TITLE
feature/Decouple MessageListAdapter from MessageListFragment

### PR DESCRIPTION
**Pull request by bhm**
I'm seeing quite a few PR's that are supposed to migrate ListView to Recycler. This is not as simple and evidently breaks in quite a few places.

This attempts, as the title indicates, to first decouple the adapter from the fragment reference.  I'm started to ride the snapshot with these changes as a daily. So far so good.

Further work is indicated by TODO's/FIXME's, but according to Code Style, they're not welcome. I'll try to clean that up later on.

Plan in further PRs is to

1. Separate ViewHolder from Adapter further.
2. Separate Cursor data extraction logic out of the adapter and viewholder.
3. Swap in the Recycler and Recycler.Adapter.
4. If possible/viable, allow for generics in the adapter, so we can have multiple object types on the MessageList view.
5. Provide a a proper Empty View/Load More to better follow the up to date design trends. (given this is the direction the project wants to go)

This in turn is a plan to refactor the monstrum that is the MessageListFragment, and finally allow for future ViewModel (from code reading this appears to be the direction) implementation for that part of code.

### CODE REVIEW

**Comment by cketti** [comment]
Sounds like a good plan :+1: Ideally it'll be implemented using small steps to make pull requests easy to review.

Unfortunately this pull request already is too big to not be painful to review. Can you please split this into multiple parts? It looks like you converted some classes to Kotlin. That's awesome :+1: But please make such changes in a separate commit. Better yet, use the [VCS Kotlin Converter](https://plugins.jetbrains.com/plugin/10862-vcs-kotlin-converter/) plugin and make it two commits. That way it'll be much easier to get the complete file history later. Also, please keep the conversion to Kotlin as simple and true to the original code as possible. If you want to improve the code, make it a separate commit.

**Comment by bhm** [comment]
rebase&squash habit got the better of me. I think reflog won't help me at this point.

Will reproduce this PR as separate commits in the following days. I'll try to keep it as granular as possible.

**Comment by bhm** [comment]
Reproduced it with a smaller scope. 

I left out the the current implementation of MessageViewHolder - will come in further PRs.

I'll push it to my phone and ride it for a couple of days  to have it tested.

**Comment by bhm** [comment]
@cketti is there a way to download artefact of this build?

File: app/ui/src/main/java/com/fsck/k9/fragment/MessageListFragment.java [hunk 1]
```diff
@@ -169,7 +169,6 @@ public static MessageListFragment newInstance(
     private MessageListAdapter adapter;
     private View footerView;
     private FolderInfoHolder currentFolder;
-    private LayoutInflater layoutInflater;
```
Comment (cketti) [inline app/ui/src/main/java/com/fsck/k9/fragment/MessageListFragment.java]: Why remove the field and look up the `LayoutInflater` every time one is needed?

File: app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt [hunk 2]
```diff
@@ -62,6 +63,9 @@ class MessageListAdapter internal constructor(
         array.recycle()
     }
 
+    private inline val previewLines: Int
```
Comment (cketti) [inline app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt]: Instead of moving around the code that is accessing settings via `K9` I'd prefer to pass in the settings. It's probably best to group all of these settings in a `MessageListAppearance` data class.

File: app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt [hunk 3]
```diff
@@ -32,7 +32,8 @@ class MessageListAdapter internal constructor(
         private val res: Resources,
         private val fragment: MessageListFragment,
         private val layoutInflater: LayoutInflater,
-        private val messageHelper: MessageHelper
+        private val messageHelper: MessageHelper,
+        private val showingThreadedList: Boolean = false
```
Comment (cketti) [inline app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt]: This should also go into the `MessageListAppearance` class.

File: app/ui/src/main/java/com/fsck/k9/fragment/MessageListFragment.java [hunk 4]
```diff
@@ -1,18 +1,6 @@
 package com.fsck.k9.fragment;
 
 
-import java.util.ArrayList;
```
Comment (cketti) [inline app/ui/src/main/java/com/fsck/k9/fragment/MessageListFragment.java]: Commit 676eeeab10884456a5f70ce708a8aa5116ecbaf7 contains a lot of unrelated code style changes. Please get rid of these.

File: app/ui/src/main/java/com/fsck/k9/fragment/MessageListFragment.java [hunk 5]
```diff
@@ -2526,10 +2526,12 @@ public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
         if (cursors.length > 1) {
             cursor = new MergeCursorWithUniqueId(cursors, getComparator());
             uniqueIdColumn = cursor.getColumnIndex("_id");
+
         } else {
             cursor = data;
             uniqueIdColumn = ID_COLUMN;
         }
+        if (adapter != null) adapter.setUniqueIdColumn(uniqueIdColumn);
```
Comment (cketti) [inline app/ui/src/main/java/com/fsck/k9/fragment/MessageListFragment.java]: A method on `adapter` is called without a `null` check further down. So this one shouldn't be necessary either.

File: app/ui/src/main/java/com/fsck/k9/fragment/AccountRetriever.kt [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+import android.database.Cursor
+import com.fsck.k9.Account
+import com.fsck.k9.Preferences
+import com.fsck.k9.fragment.MLFProjectionInfo.ACCOUNT_UUID_COLUMN
+
+class AccountRetriever constructor(
+        private val preferences: Preferences
+) : (Cursor) -> Account {
+
+    override fun invoke(cursor: Cursor): Account {
```
Comment (cketti) [inline app/ui/src/main/java/com/fsck/k9/fragment/AccountRetriever.kt]: In my opinion the operator overloading doesn't make the code easier to read. On the contrary, it's harder to figure out what's going on. Especially for people coming from Java.

Also, it feels like this is a helper function at best. The implementation is highly coupled to the used projection. So it probably shouldn't live in its own class.

File: app/ui/src/test/java/com/fsck/k9/fragment/AccountRetrieverTest.kt [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import org.mockito.BDDMockito.given
+import org.mockito.Mockito.mock
+
+class AccountRetrieverTest {
+
+    @Test fun shouldRetrieveAccountCorrectly() {
+        // given
+        val someUuid = "someUuid";
+        val expectedAccount = Account(someUuid)
+        val preferences = mock(Preferences::class.java)
```
Comment (cketti) [inline app/ui/src/test/java/com/fsck/k9/fragment/AccountRetrieverTest.kt]: `mockito-kotlin` makes this nicer to read (and write). See e.g. https://github.com/k9mail/k-9/blob/f73ef35892d5085f3ffde109563cdae07da7578c/app/core/src/test/java/com/fsck/k9/message/quote/TextQuoteCreatorTest.kt#L24-L33

File: app/ui/src/test/java/com/fsck/k9/fragment/AccountRetrieverTest.kt [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+import org.junit.Test
+import org.mockito.ArgumentMatchers.anyInt
+import org.mockito.ArgumentMatchers.anyString
+import org.mockito.BDDMockito.given
+import org.mockito.Mockito.mock
+
+class AccountRetrieverTest {
+
+    @Test fun shouldRetrieveAccountCorrectly() {
+        // given
```
Comment (cketti) [inline app/ui/src/test/java/com/fsck/k9/fragment/AccountRetrieverTest.kt]: We use empty lines between the 'given', 'when', and 'then' blocks. That way we don't need comments.

File: app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt [hunk 9]
```diff
@@ -33,11 +33,12 @@ class MessageListAdapter internal constructor(
         context: Context,
         theme: Resources.Theme,
         private val res: Resources,
-        private val fragment: MessageListFragment,
         private val layoutInflater: LayoutInflater,
         private val messageHelper: MessageHelper,
         private val contactsPictureLoader: ContactPictureLoader,
         private val accountRetriever: AccountRetriever,
+        private val toggleMessageSelectWithAdapterPosition: (Int) -> Unit,
```
Comment (cketti) [inline app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt]: As you probably have noticed, using a function type like this makes it very ugly to use from Java. I think I'd prefer to have all callbacks in an interface. Something like `MessageListItemActionListener`.

File: app/ui/src/main/java/com/fsck/k9/fragment/MessageListFragment.java (hunk 1 above)
Comment (bhm) [reply app/ui/src/main/java/com/fsck/k9/fragment/MessageListFragment.java]: I'll look it up again, but I believe the only reason for this field and a getter exit is for the list adapter.

I'll leave it as a field if we do  another service location.

File: app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt (hunk 2 above)
Comment (bhm) [reply app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt]: Yep, thought about this during last commute. I'll try to pack it up. Down the line this could allow for adapter refresh on settings change.

File: app/ui/src/main/java/com/fsck/k9/fragment/AccountRetriever.kt (hunk 6 above)
Comment (bhm) [reply app/ui/src/main/java/com/fsck/k9/fragment/AccountRetriever.kt]: Removed, agreed. Data unpacking logic will get moved out of Adapter/ViewHolder at some point. 

Class was a poor attempt at removing the fragment dependency, which was evident once I started again at ViewHolder refactor.

File: app/ui/src/test/java/com/fsck/k9/fragment/AccountRetrieverTest.kt (hunk 7 above)
Comment (bhm) [reply app/ui/src/test/java/com/fsck/k9/fragment/AccountRetrieverTest.kt]: Outdated, tested class was removed.

File: app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt (hunk 9 above)
Comment (bhm) [reply app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt]: Done in a separate commit.

File: app/ui/src/test/java/com/fsck/k9/fragment/AccountRetrieverTest.kt (hunk 8 above)
Comment (bhm) [reply app/ui/src/test/java/com/fsck/k9/fragment/AccountRetrieverTest.kt]: Outdated for now, but will remember for the future.

**Comment by bhm** [comment]
This PR fails at some license check during JDK install.

**Comment by bhm** [comment]
Part of the issues got addressed as separate commits. Feel free to mark comments resolved.

I'll address the rest during this week.

File: app/ui/src/main/java/com/fsck/k9/fragment/MessageListFragment.java (hunk 5 above)
Comment (bhm) [reply app/ui/src/main/java/com/fsck/k9/fragment/MessageListFragment.java]: done

File: app/ui/src/main/java/com/fsck/k9/fragment/MessageListFragment.java (hunk 4 above)
Comment (bhm) [reply app/ui/src/main/java/com/fsck/k9/fragment/MessageListFragment.java]: done

File: app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt (hunk 3 above)
Comment (bhm) [reply app/ui/src/main/java/com/fsck/k9/fragment/MessageListAdapter.kt]: done.

**Comment by cketti** [comment]
I've cleaned up the commit history and created PR #4140. I think it's ready to be merged. But I'd appreciate a quick look.
